### PR TITLE
Overmap overhaul (step 4)

### LINKED ISSUE #19414: overmap specials rotation only works for north facing terrains
Example:
If I have a declaration like this in an overmap_specials.json:
```
"overmaps" : [
      { "point":[0,0,0], "overmap": "house_west"},
      { "point":[1,0,0], "overmap": "house_north"},
      { "point":[2,0,0], "overmap": "house_east"}
]
```
then I get over terrain with correct facings if the whole structure faces North
```
<^>
```
but if it is randomly rotated to the East then every terrain faces East
```
>
>
>
```
when it should be
```
^
>
v
```
Win10, Tiles, 0.C-19846-gebe6fa5

### LINKED ISSUE #19508: "Mapgen" in Debug Menu Map Editor crashes game
Build 5966/Version 0.C-19952-g3d4a0eb (tiles)

Opening the debug menu, going to the Map Editor, and selecting Mapgen causes a crash,

### PR DESCRIPTION
**Pull request by codemime**
Continues #19223.

Fixes #19414.
Fixes #19508.

### Observable changes

~~The whole thing is presumably faster due to less number of string comparisons in nested loops. A benchmark will follow.~~ 

EDIT: Average values after generating 30 overmaps:

Function			| Calls per overmap | Before (ms) | After (ms) 
--|--|--|--
`generate()`		| 1 | 334 | 255
`polish(0)`		| 1 | 36  | 13
`polish(z, "ants")`		| ~6 | 4 | 1
`polish(z, "sewer")`	| ~6 | 4 | 1
`polish(z, "subway")`	| ~6 | 4 | 1

Time spent on generating has decreased to 76% of the previous. It's 1,3 times faster now (and it's not the limit)!

### Implementation details

- The code that works with overmap terrains became much less spaghettified (for example, `overmap::good_road()` was a long chain of `if-else` statements, now it's ten times shorter. 
See c59ef66).
- Overmap lines now have a dedicated namespace and structures (see  7e50eea).
- Immutable data from `oter_t` was extracted into a separate "flyweight" class `oter_type_t`. `oter_t` delegates its functions to it (encapsulated with `get_foo()`). See d069ed1.
- Overmap terrains and their types are loaded using `generic_factory`. This (along with the extraction) reduces overall memory consumption. Prior to this PR:
    - Each `oter_t` was duplicated (one in `oterlist` and another one in `otermap`).
    - Each `oter_t` was duplicated 3 (rotatable) and 15 (linear) times. This gives 6 and 30 dupes respectively.
    - Now they still have 3 and 15 variations, but keep only distinctive data.
- Redundant fields were removed from `overmap_terrain.json` files. Haven't touched the "flag" values yet: too much for one PR.
- Added dead-ends to subways, sewers, and ant tunnels.
- Overmap terrain symbols are rotated after loading (instead of loading duplicated sequences `[^,>,v,<]`). Only one symbol per terrain now:
    - No need to bloat JSON with repetitive stuff (a source of errors).
    - `"rotate"` and `"sym"` fields were coupled. We had to update them only together.
    - Each terrain object is quite isolated now (doesn't know anything about the others).
- Overmap weighted lists now use terrain types.
- Overmap spawns were slightly de-duplicated (see 99c09e4).
- String comparisons in finding a suitable location (overmap specials) and "polishing" (overmap terrains) were mitigated (see 69abf53 and d698de9).
- `int_id`s now have a dereference operator `*`.
    
Sorry for the somewhat big PR. Yet nearly a half of its size is pretty trivial JSON / encapsulation changes.

### Changes to JSON API

`sym` in `overmap_terrain` now accepts only one symbol (e.g. `"sym": 94` instead of `"sym": [ 94, 62, 118, 60 ]`). The symbol still rotates accordingly if `"rotate": true` is specified.
Currently rotatable symbols are: 
```
     60, 62, 94, 118,
     4194410,
     4194411,
     4194412,
     4194413,
     4194417,
     4194421,
     4194422,
     4194423,
     4194424.
```
More symbols may be added on request. Changed in #19518.

### CODE REVIEW

**Comment by mugling** [comment]
> Only one symbol per terrain now

You need to update linter rule for `overmap_terrain:sym` - it doesn't need `NOWRAP` any more

**Comment by mugling** [comment]
Can you write a small summary of API changes to go in the relevant thread and post either as a comment or in your preamble

**Comment by codemime** [comment]
The change of symbols is the only API change (apart from strict JSON fields).

**Comment by mugling** [comment]
> The change of symbols is the only API change (apart from strict JSON fields).

I would suspect so. Can you post something to this thread in the appropriate format so when it's merged we can add an entry to the JSON API notifications. I'd prefer if you could post there yourself but that is as-yet not possible.

**Comment by codemime** [comment]
> Can you post something to this thread in the appropriate format so when it's merged we can add an entry to the JSON API notifications.

Added to the description.

File: src/mapgen.cpp [hunk 1]
```diff
@@ -3279,26 +3278,32 @@ C..C..C...|hhh|#########\n\
                 rotate(3);
             }
         } else if (tw != 0 || rw != 0 || lw != 0 || bw != 0) { // Sewers!
+
+            const auto is_lab = []( const oter_id &id ) {
```
Comment (mugling) [inline src/mapgen.cpp]: Can you document this? I suspect your trying to stop the lab cutting into sewers?

File: src/mapgen_functions.cpp [hunk 2]
```diff
@@ -175,16 +175,19 @@ void init_mapgen_builtin_functions() {
 
     mapgen_cfunction_map["subway_straight"]    = &mapgen_subway_straight;
     mapgen_cfunction_map["subway_curved"]      = &mapgen_subway_curved;
+    mapgen_cfunction_map["subway_end"]         = &mapgen_subway_straight;   // @todo Implement
```
Comment (mugling) [inline src/mapgen_functions.cpp]: Why todo?

**Comment by Coolthulhu** [comment]
Has merge conflict

File: src/mapgen_functions.cpp (hunk 2 above)
Comment (codemime) [inline src/mapgen_functions.cpp]: They're not really dead-ends (though they could be), they simply duplicate straight segments to eliminate the special case. I'll document this better.

File: src/mapgen.cpp (hunk 1 above)
Comment (codemime) [inline src/mapgen.cpp]: No, not really. I was getting rid of unreliable operators (see eb653822d4264fc35809f3a00fc46212082c964f). 
#14136 is a separate issue, I didn't mean mending it here (to stick with the pure refactoring).

**Comment by codemime** [comment]
Rebased and added a benchmark. All set.

**Comment by mugling** [comment]
Try and avoid rebasing as it requires the reviewer to read all of your code again to find out what changed

**Comment by codemime** [comment]
Noted.

**Comment by mugling** [comment]
Can you post your summary to the API changes thread in roughly the same format as the existing messages

**Pull request merged**